Feature/search page and autocomplete

codeAdrian opened the pull request:
<!-- (REQUIRED) What is the nature of this PR? -->

## This PR is a:

[x] New feature
[ ] Enhancement/Optimization
[ ] Refactor
[ ] Bugfix
[ ] Test for existing code
[ ] Documentation

<!-- (REQUIRED) What does this PR change? -->

## Summary

![screen shot 2018-11-27 at 11 05 28 am](https://user-images.githubusercontent.com/11479290/49074211-834ec980-f234-11e8-9c58-9180a77eadd1.png)

When this pull request is merged, it will add header search autocomplete component. Autocomplete component returns category and product suggestions and handles click actions on those links.

Related issue: https://github.com/magento-research/pwa-studio/issues/313

**Note:** I'm discussing some minor UX improvements to this branch with Soumya, I will leave a comment and update PR when we see about the improvements, but we can proceed with code review and see if anything needs to be changed.

<!-- (OPTIONAL) What other information can you provide about this PR? -->

## Additional information

* This feature depends on Crifen's "search-results" branch. It has been updated with their latest changes. After their latest update, the routing to product page doesn't work on search results page (on other pages, it work as expected). I assume that this is still work in progress on their side.
* Edited Crifen's searchBar component to accommodate the autocomplete component: added additional event handlers, replaced refs with state where needed, moved search query to separate query file since it will be shared between autocomplete and search component, and some other minor changes.
* Added query-string npm package to "venia-concept" package.json in order to better handle query strings from URL and Router (used on search Root component, improvement on Crifen's code)
* Typing event in autocomplete is debounced so that the query runs when user stopped typing for 200ms.
* Suggested categories are limited to 4, suggested products are limited to 3.


<!--
Thank you for your contribution!

Before submitting this pull request, please make sure you have read our Contribution Guidelines and your PR meets our contribution standards:
https://github.com/magento-research/pwa-studio/blob/master/.github/CONTRIBUTION.md

Please fill out as much information as you can about your PR to help speed up the review process.
If your PR addresses an existing GitHub Issue, please refer to it in the title or Additional Information section to make the connection.

We may ask you for changes in your PR in order to meet the standards set in our Contribution Guidelines. PR's that do not comply with our guidelines may be closed at the maintainers' discretion.

Feel free to remove this section before creating this PR.
-->

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento-research/pwa-studio?pullRequest=548) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento-research/pwa-studio?pullRequest=548) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: Crifen<br/>:x: codeAdrian

codeAdrian commented:
Hello everyone, I am detailing my updates for tests in Crifen's searchBar component.

* 'When isOpen is true, the input component is focused.' - this test was removed because this functionality has been removed. When user is redirected to search results page after search, the input remains focused and keyboard on mobile device will activate, hiding part of the website. This can affect the UX in negative way. Assuming that if user has performed the search, they would usually like to preview the results before entering another search query.

* 'When isOpen is false, the input is blurred.' - same edit as above

* ‘When the search bar is expanded, pressing the Enter key will submit.’ - edited the test to accomodate the change of adding <form> element and onsubmit event.

* 'When the search icon is clicked, the query in the input component will be submitted.' - same edit as above.

* ‘When the input component is empty, pressing the enter key will not search.’ - same edit as above.

* 'When url is pointed to search results page, the search input will get its value from the url.' - adjusted test to use React Router instead of window.location object

codeAdrian commented:
Hello everyone,
I have updated this PR with Crifen's latest changes. I have also included UX improvement which Soumya suggested. We are still considering one minor UX improvement. I will keep you posted about that.

codeAdrian commented:
> Nice demo today, @codeAdrian! For further improvements, I would suggest modifying the logic for how the suggested product categories are calculated (currently based on the first page only). Ideally, we'd take the entire result set of products, look at the total set of their category assignments, and return the top most popular categories from that list.
> There is actually a way to look at the layered navigation filters via graphQL by querying the `filters` object, but it currently has some issues like [magento/graphql-ce#258](https://github.com/magento/graphql-ce/issues/258). The purpose of that endpoint is to return the layered nav filters that you would see on storefront (both in case of browsing and searching).
> cc: @jimbo

Hi @misha-kotov , glad you enjoyed the demo and thank you for your suggestion. I will look into the GraphQL for filters object and see what I can do and report back.

codeAdrian commented:
Hello @misha-kotov @ericerway , I have taken a look at the improvement that @misha-kotov has suggested. As Misha has said, GraphQL implementation of layered nav filters has some issues. Here's what I have found:

I have simulated the search query in GraphiQL and compared it to same search query in Luma theme in Magento 2.3. Currently, the GraphQL implementation of filters do not reflect the actual state of filters in search. Compare the following screenshots:

GraphiQL simulation of search query with filters returned.

![screen shot 2018-12-04 at 8 30 30 am](https://user-images.githubusercontent.com/11479290/49427008-800e8d00-f7a2-11e8-9310-27ba28e061fb.png)

Magento 2.3. Luma theme search.

![screen shot 2018-12-04 at 8 46 03 am](https://user-images.githubusercontent.com/11479290/49427021-843aaa80-f7a2-11e8-8037-8ca8e78fcf1b.png)

Notice how the "items_count" variable in each filter is different than the numbers next to each filters in Luma. For example, in GraphQL returned count for "Bottoms" is 24 which is incorrect (total item count is 8), while in Luma returned count is 7 which is correct. These results never change in GraphQL implementation (it always returns 24 for "Bottoms", for example), regardless of the query (search, filter by category, etc.).

Another limitation is that only the topmost categories are covered by the filters, while the current implementation covers both top categories and subcategories.

Let me know if I should proceed with the Misha's suggestion and assume that the GraphQL layered nav issues will be fixed (and also limit the filter to topmost categories) or if we should keep the current implementation.

codeAdrian commented:
Hello everyone, Pull Request has been updated with the following changes:
* Merged latest changes from @supernova-at and made minor adjustments for autocomplete to work properly
* Added final UX improvements which were discussed with @soumya-ashok 
* Changed GraphQL query per advice from @misha-kotov 
* Adjusted tests to accommodate for changes in code 
* Code cleanup and minor fixes

Let me know if I need to make any changes

supernova-at commented:
> I'm running into issues when running "yarn test" command (both on this branch and release branch).

Hmm, I'm getting a completely different test failure on your branch:

```
 FAIL  packages/venia-concept/src/components/SearchBar/__tests__/searchBar.spec.js
  ● entering text in the input causes the reset button to appear

    TypeError: Cannot read property 'value' of undefined

      77 |     // Simulate entering text in the input field.
      78 |     const input = instance.find(inputTypes);
    > 79 |     input.props.onChange({
         |                 ^
      80 |         currentTarget: { value: 'some text' }
      81 |     });
      82 |
```

I am not getting any failures when running against the release branch 🤔

jimbo commented:
@supernova-at I'm seeing exactly what you're seeing.

@codeAdrian have you cleared out your packages? `npm run clean:all`, `npm install`, `npm run build`

jimbo commented:
I've resolved the broken test that @supernova-at and I were seeing. CI is passing now as well.

@codeAdrian I think the issues on your end could be related to something else. Try clearing out your packages as I suggested above and see if that helps. It could also be related to yarn; as of now, we assume people are using npm.

codeAdrian commented:
@jimbo @supernova-at Thank you for looking into this and fixing the test. I assumed that something was up with the latest release since I never got this issue beforehand. I will run the clean command and use npm instead of yarn and let you know tomorrow if the issue has been resolved.

codeAdrian commented on packages/peregrine/src/index.js:
```diff
@@ -5,6 +5,6 @@ export { default as ContainerChild } from './ContainerChild';
 export { default as Page } from './Page';
 export { default as Price } from './Price';
-export { default as Router } from './Router';
+export { default as Router, RouteConsumer } from './Router';
```
For some reason, RouteConsumer changes weren't properly merged so I fixed them manually. Might have been caused by multiple rebases. In any case, I have fixed this. Let me know if I have missed anything else.

codeAdrian commented on packages/venia-concept/src/RootComponents/Search/search.js:
```diff
@@ -46,44 +37,75 @@ export class Search extends Component {
         toggleSearch: func
     };
 
-    componentDidMount() {
```
Thank you for noticing this. This functionality was lost by the latest release merge, I have restored this functionality.

codeAdrian commented on packages/venia-concept/src/RootComponents/Search/search.js:
```diff
@@ -94,4 +116,13 @@ export class Search extends Component {
+const mapDispatchToProps = { executeSearch };
+
+export default compose(
+    withRouter,
```
Fixed by removing withRouter from the component.

codeAdrian commented on packages/venia-concept/src/components/Header/header.css:
```diff
@@ -5,7 +5,21 @@
     grid-auto-columns: 100%;
     grid-auto-flow: row;
     grid-auto-rows: auto;
+    transition-property: grid-row-gap;
```
This was left over by the merge. I have manually updated the file. It should be same as on the release.

codeAdrian commented on packages/venia-concept/src/components/SearchBar/searchBar.js:
```diff
@@ -42,24 +61,54 @@ export class SearchBar extends Component {
-    handleSubmit = ({ search_query }) => {
-        this.props.executeSearch(search_query, this.props.history);
+    handleSubmit = () => {
+        const { searchQuery } = this.state;
```
Removed the searchQuery state and replaced it with Form API state. handleChange was necessary for the input to keep track of user typing so the Autocomplete component is activated.

codeAdrian commented on packages/venia-concept/src/components/SearchBar/suggestedCategories.js:
```diff
@@ -0,0 +1,54 @@
+                render="ul"
+                className={classes.root}
+                items={categorySuggestions}
+                getItemKey={item => item['value_string']}
```
Fixed this by applying the suggested change.

codeAdrian commented on packages/venia-concept/src/components/SearchBar/suggestedProduct.css:
```diff
@@ -0,0 +1,26 @@
+.root {
+    position: relative;
```
Leftover from the initial styling. This line has been removed.

codeAdrian commented:
Hello @jimbo , thank you for taking the time to check everything everything and note the changes. I have fixed everything you have suggested, let me know if you have any questions or if I need to make any additional changes. Also, the team can feel free to make any minor and simple adjustments, so you don't have to wait for me to do it the next day (due to the timezone differences).

Regarding the issues with testing, running the clean command, running install and using npm instead of yarn has fixed it. Thank you for the assistance.

jimbo commented:
@codeAdrian Thanks for making those changes! I'll sweep through and make a couple of quick changes, then we'll go ahead and merge this.

jimbo commented:
Okay, I've made my changes and stripped the lockfile changes from this PR. 113 commits later, I think we're ready to go with this.

@supernova-at Speak now or forever hold your peace!

supernova-at commented:
> @supernova-at Speak now or forever hold your peace!

I'm cool with it - the autocomplete popup doesn't go away upon executing a search, but I'm fine creating a follow up bug ticket for that.

jimbo commented:
@codeAdrian Merged! Thanks again for sticking with this one. 😅
